share/CMakeJS.cmake

```
@@ -0,0 +1,642 @@
+	file(GLOB NODE_ADDON_API_INC_FILES "${NODE_ADDON_API_DIR}/*.h")
+	set(NODE_ADDON_API_INC_FILES "${NODE_ADDON_API_INC_FILES}" CACHE FILEPATH "Node Addon API Header files." FORCE)
+        source_group("Node Addon API (C++)" FILES "${NODE_ADDON_API_INC_FILES}")
+    endif()
+endif()
+
+#[=============================================================================[
+Silently create an interface library (no output) with all Addon API dependencies
+resolved, for Addon targets to link with.
+
+(This should contain most of cmake-js globally-required configuration)
+
+Targets:
+
+cmake-js::node-dev
+cmake-js::node-api
+cmake-js::node-addon-api
+cmake-js::cmake-js
+
+]=============================================================================]#
+
+# NodeJS system installation headers
+# cmake-js::node-dev
+add_library                 (node-dev INTERFACE)
+add_library                 (cmake-js::node-dev ALIAS node-dev)
+target_include_directories  (node-dev INTERFACE "${CMAKE_JS_INC}")
+target_sources              (node-dev INTERFACE "${CMAKE_JS_SRC}")
+target_sources              (node-dev INTERFACE "${CMAKE_JS_INC_FILES}")
+target_link_libraries       (node-dev INTERFACE "${CMAKE_JS_LIB}")
+
+# Node API (C) - requires NodeJS system installation headers
+# cmake-js::node-api
+add_library                 (node-api INTERFACE)
+add_library                 (cmake-js::node-api ALIAS node-api)
+target_include_directories  (node-api INTERFACE "${NODE_API_HEADERS_DIR}")
+target_sources              (node-api INTERFACE "${NODE_API_INC_FILES}")
+target_link_libraries       (node-api INTERFACE cmake-js::node-dev)
+
+# Node Addon API (C++) - requires Node API (C)
+# cmake-js::node-addon-api
+add_library                 (node-addon-api INTERFACE)
+add_library                 (cmake-js::node-addon-api ALIAS node-addon-api)
+target_include_directories  (node-addon-api INTERFACE "${NODE_ADDON_API_DIR}")
+target_sources              (node-addon-api INTERFACE "${NODE_ADDON_API_INC_FILES}")
+target_link_libraries       (node-addon-api INTERFACE cmake-js::node-api)
+
+# CMakeJS API - requires Node Addon API (C++), resolves the full Napi Addon dependency chain
+# cmake-js::cmake-js
+add_library                 (cmake-js INTERFACE)
+add_library                 (cmake-js::cmake-js ALIAS cmake-js)
+target_link_libraries       (cmake-js INTERFACE cmake-js::node-addon-api)
+target_compile_definitions  (cmake-js INTERFACE "BUILDING_NODE_EXTENSION")
+target_compile_features     (cmake-js INTERFACE cxx_nullptr) # Signal a basic C++11 feature to require C++11.
+
+# Generate definitions
+if(MSVC)
+    set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>" CACHE STRING "Select the MSVC runtime library for use by compilers targeting the MSVC ABI." FORCE)
+    if(CMAKE_JS_NODELIB_DEF AND CMAKE_JS_NODELIB_TARGET)
+        execute_process(COMMAND ${CMAKE_AR} /def:${CMAKE_JS_NODELIB_DEF} /out:${CMAKE_JS_NODELIB_TARGET} ${CMAKE_STATIC_LINKER_FLAGS})
+    endif()
```

```suggestion
    set(CMAKE_JS_NODELIB_TARGET "${CMAKE_BINARY_DIR}/node.lib")
    if(CMAKE_JS_NODELIB_DEF)
        execute_process(COMMAND ${CMAKE_AR} /def:${CMAKE_JS_NODELIB_DEF} /out:${CMAKE_JS_NODELIB_TARGET} ${CMAKE_STATIC_LINKER_FLAGS})
        target_link_libraries       (node-dev INTERFACE "${CMAKE_JS_NODELIB_TARGET}")
    endif()
```

This should probably be moved, as this also replaces the line `target_link_libraries       (node-dev INTERFACE "${CMAKE_JS_LIB}")` and any usages of `CMAKE_JS_LIB`

---

Also `NODE_API_DEF_PATH` should be auto-populated with something like the following
```
execute_process(COMMAND ${NODE_PATH} -p "require('node-api-headers').def_paths.node_api_def"
    WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
    OUTPUT_VARIABLE NODE_API_DEF_PATH
    OUTPUT_STRIP_TRAILING_WHITESPACE
)
```

---

> Also `CMAKEJS_NODELIB_DEF` should be auto-populated with something like the following
> 
> ```
> execute_process(COMMAND ${NODE_PATH} -p "require('node-api-headers').def_paths.node_api_def"
>     WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
>     OUTPUT_VARIABLE CMAKEJS_NODELIB_DEF
>     OUTPUT_STRIP_TRAILING_WHITESPACE
> )
> ```

Yeah I never quite "got" this. I do have a win 11 x64 I build stuff against, but - my next major sugggestion request for cmake-js - I dislike the Visual Studio dependency cmake-js has on windows platforms! Not that I dislike Visual Studio or MSBuildtools - far from it - but I just happen to be a huge fan of MSYS2 (and cygwin-like things). Well spotted!

These sorts of things can actually also be made into targets using ```add_custom_target()``` . the user may do a build run specifying a unique target and only that target will be built ('clean' and 'install' are both actually just custom targets you get given by default).

It's possible to make a custom target for *every* part of the dependency chain, potentially allowing the user full control of how they want to build programmatically:

```.cmake
add_custom_target(install_node_modules <some_npm_install_stuff>)
```

I suppose that if the above-suggested target(s) were defined *instead of* all those ```execute_process()```, then everything will still run in the correct order.

---

> Yeah I never quite "got" this. I do have a win 11 x64 I build stuff against, but - my next major sugggestion request for cmake-js - I dislike the Visual Studio dependency cmake-js has on windows platforms! Not that I dislike Visual Studio or MSBuildtools - far from it - but I just happen to be a huge fan of MSYS2 (and cygwin-like things). Well spotted!

haha, I try to avoid windows whenever possible, so havent minded that this can only do msvc, but yeah it being able to find msys2 and cygwin would be reasonable. It would probably be pretty trivial to make this new file support it, with a bit of js needed for the 'magic' stuff to tell cmake the correct generator, where it is choosing a sensible vs version currently.

> These sorts of things can actually also be made into targets using add_custom_target() 

I think that is a good plan for many things.  
Im not sure if `generate_windows_defs` should be done that way though, as that does need to be done for MSVC to build, so it would be better to do automatically.
And im not sure if `install_node_modules` should be one either, as it could need to execute any of yarn, npm or pnpm and also needs to be run before the `require('node-addon-api')` can be performed, which I suspect will make ensuring correct order hard. So I feel like this would be better to simply be some documentation telling users to put an execute_process at the top of their cmakelists file instead